crates/api/src/linker.rs

```diff
@@ -0,0 +1,368 @@
+///
+/// As mentioned previously, `Linker` is a form of name resolver. It will be
+/// using the string-based names of imports on a module to attempt to select a
+/// matching item to hook up to it. This name resolution has two-levels of
+/// namespaces, a module level and a name level. Each item is defined within a
+/// module and then has its own name. This basically follows the wasm standard
+/// for modularization.
+///
+/// Names in a `Linker` can be defined twice, but only for different signatures
+/// of items. This means that every item defined in a `Linker` has a unique
+/// name/type pair. For example you can define two functions with the module
+/// name `foo` and item name `bar`, so long as they have different function
+/// signatures. Currently duplicate memories and tables are not allowed, only
+/// one-per-name is allowed.
+pub struct Linker {
+    store: Store,
+    string2idx: HashMap<Rc<str>, usize>,
+    strings: Vec<Rc<str>>,
+    map: HashMap<ImportKey, Extern>,
+}
+
+#[derive(Hash, PartialEq, Eq)]
+struct ImportKey {
+    name: usize,
+    module: usize,
+    kind: ImportKind,
+}
+
+#[derive(Hash, PartialEq, Eq, Debug)]
+enum ImportKind {
+    Func(FuncType),
+    Global(GlobalType),
+    Memory,
+    Table,
+}
+
+impl Linker {
+    /// Creates a new [`Linker`].
+    ///
+    /// This function will create a new [`Linker`] which is ready to start
+    /// linking modules. All items defined in this linker and produced by this
+    /// linker will be connected with `store` and must come from the same
+    /// `store`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use wasmtime::{Linker, Store};
+    ///
+    /// let store = Store::default();
+    /// let mut linker = Linker::new(&store);
+    /// // ...
+    /// ```
+    pub fn new(store: &Store) -> Linker {
+        Linker {
+            store: store.clone(),
+            map: HashMap::new(),
+            string2idx: HashMap::new(),
+            strings: Vec::new(),
+        }
+    }
+
+    /// Defines a new item in this [`Linker`].
+    ///
+    /// This method will add a new definition, by name, to this instance of
+    /// [`Linker`]. The `module` and `name` provided are what to name the
+    /// `item`.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the `module` and `name` already identify an item
+    /// of the same type as the `item` provided. For more information see the
+    /// documentation on [`Linker`].
+    ///
+    /// Also returns an error if `item` comes from a different store than this
+    /// [`Linker`] was created with.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use wasmtime::*;
+    /// # fn main() -> anyhow::Result<()> {
+    /// # let store = Store::default();
+    /// let mut linker = Linker::new(&store);
+    /// let ty = GlobalType::new(ValType::I32, Mutability::Const);
+    /// let global = Global::new(&store, ty, Val::I32(0x1234))?;
+    /// linker.define("host", "offset", global);
```

Comment:
This should have a `?` to handle errors.

Reply:
Oh good eye, I've added a commit to https://github.com/bytecodealliance/wasmtime/pull/1391 which denies warnings in doctests to catch this